python_scripts/linear_models_feature_engineering_classification.py

```diff
@@ -0,0 +1,437 @@
+# assigned to the class 1 while the points in the outer edges are assigned to
+# the class 0.
+
+# %%
+from sklearn.datasets import make_gaussian_quantiles
+
+X, y = make_gaussian_quantiles(
+    n_samples=100, n_features=2, n_classes=2, random_state=42
+)
+gauss = pd.DataFrame(
+    np.concatenate([X, y[:, np.newaxis]], axis=1),
+    columns=feature_names + [target_name],
+)
+data_gauss, target_gauss = gauss[feature_names], gauss[target_name]
+
+# %% [markdown]
+#
+# The third dataset is called the "XOR" dataset as the data points are sampled
+# from a uniform distribution in a 2D space and the class is defined by the
+# Exclusive OR (XOR) operation on the two features: the target class is 1 if
+# only one of the two features is greater than 0. The target class is 0
+# otherwise.
+
+# %%
+xor = pd.DataFrame(
+    np.random.RandomState(0).uniform(low=-1, high=1, size=(200, 2)),
+    columns=feature_names,
+)
+target_xor = np.logical_xor(xor["Feature #0"] > 0, xor["Feature #1"] > 0)
+target_xor = target_xor.astype(np.int32)
+xor["class"] = target_xor
+data_xor = xor[feature_names]
+
+# %% [markdown]
+#
+# We use matplotlib to visualized the data points in both datasets:
+
+# %%
+import matplotlib.pyplot as plt
+import seaborn as sns
+import warnings
+
+
+warnings.filterwarnings("ignore", module="seaborn")
+_, axs = plt.subplots(ncols=3, figsize=(14, 4))
+
+sns.scatterplot(
+    data=moons,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_moons,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[0],
+)
+sns.scatterplot(
+    data=gauss,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_gauss,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[1],
+)
+sns.scatterplot(
+    data=xor,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_xor,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[2],
+)
+axs[0].set_title("The moons dataset")
+_ = axs[1].set_title("The Gaussian quantiles dataset")
+_ = axs[2].set_title("The XOR dataset")
+
+
+# %% [markdown]
+#
+# We intuitively observe that for each dataset, the two classes are not
+# linearly separable as the classes fold around each other in a way no straight
+# line can separate the two classes.
+#
+# We can confirm this by fitting a linear model, such as logistic regression,
+# to each dataset and plot the decision boundary of the model. Let's first define
+# a helper function to plot the decision boundary of a model:
+
+# %%
+from sklearn.inspection import DecisionBoundaryDisplay
+
+
+def plot_decision_boundary(model):
+    _, axs = plt.subplots(ncols=3, figsize=(14, 4))
+
+    for ax, (data, target) in zip(
+        axs,
+        [
+            (data_moons, target_moons),
+            (data_gauss, target_gauss),
+            (data_xor, target_xor),
+        ],
+    ):
+        model.fit(data, target)
+        DecisionBoundaryDisplay.from_estimator(
+            model,
+            data,
+            response_method="predict_proba",
```

Comment:
We introduce the concept of `predict_proba` [here in M5](https://inria.github.io/scikit-learn-mooc/python_scripts/trees_classification.html?highlight=predict_proba). In my opinion it would be better to keep the style for the decision boundary display as introduced [at the beginning of M4](https://inria.github.io/scikit-learn-mooc/python_scripts/logistic_regression.html), to rather have the discussion focused on the kernels.

Reply:
I think we can just add a quick note to explain that the background color of this plot represents the confidence level for each class (white meaning "unsure about which class to choose") using the `predict_proba` method and that more details about the use and the meaning of the `predict_proba` method will be given in the next module.

Reply:
I think this is really important to show the impact of the regularization on the confidence of the model (in the following exercise) and therefore I think it's important to introduce this way of plotting the decision boundaries as soon as possible (before the exercise, not to introduce too many new concept at once in the exercise itself).